bin/reth/src/node/mod.rs

```diff
@@ -147,65 +143,49 @@ impl Command {
+                    config.stages.bodies.downloader_max_concurrent_requests,
+            )
+            .build(fetch_client.clone(), consensus.clone(), db.clone());
+        let mut pipeline = Pipeline::builder()
             .with_sync_state_updater(network.clone())
-            .with_channel(sender)
-            .push(HeaderStage {
-                downloader: headers::linear::LinearDownloadBuilder::default()
-                    .request_limit(config.stages.headers.downloader_batch_size)
-                    .stream_batch_size(config.stages.headers.commit_threshold as usize)
-                    // NOTE: the head and target will be set from inside the stage before the
-                    // downloader is called
-                    .build(
-                        consensus.clone(),
-                        fetch_client.clone(),
-                        Default::default(),
-                        Default::default(),
-                    ),
-                consensus: consensus.clone(),
-                sync_status_updates: network.clone(),
-                metrics: HeaderMetrics::default(),
-            })
-            .push(TotalDifficultyStage {
-                commit_threshold: config.stages.total_difficulty.commit_threshold,
-            })
-            .push(BodyStage {
-                downloader: bodies::concurrent::ConcurrentDownloaderBuilder::default()
-                    .with_stream_batch_size(config.stages.bodies.downloader_stream_batch_size)
-                    .with_request_limit(config.stages.bodies.downloader_request_limit)
-                    .with_max_buffered_responses(
-                        config.stages.bodies.downloader_max_buffered_responses,
-                    )
-                    .with_concurrent_requests_range(
-                        config.stages.bodies.downloader_min_concurrent_requests..=
-                            config.stages.bodies.downloader_max_concurrent_requests,
-                    )
-                    .build(fetch_client.clone(), consensus.clone(), db.clone()),
-                consensus: consensus.clone(),
-            })
-            .push(SenderRecoveryStage {
-                batch_size: config.stages.sender_recovery.batch_size,
-                commit_threshold: config.stages.sender_recovery.commit_threshold,
-            })
-            .push(ExecutionStage {
-                chain_spec: self.chain.clone(),
-                commit_threshold: config.stages.execution.commit_threshold,
-            })
-            // This Merkle stage is used only on unwind
-            .push(MerkleStage { is_execute: false })
-            .push(AccountHashingStage { clean_threshold: 500_000, commit_threshold: 100_000 })
-            .push(StorageHashingStage { clean_threshold: 500_000, commit_threshold: 100_000 })
-            // This merkle stage is used only for execute
-            .push(MerkleStage { is_execute: true });
+            .add_stages(
```

Comment:
Uses the new stage sets: as mentioned, I want to make either the stage sets *or* the stages optionally ser/de to replace `StagesConfig`. This would make this a lot easier. Primary hurdles are:

1. The downloaders, but it is possible to make them serializable too.
2. Consensus. This should be determined by the chain spec, so should end up being ser/de too.

Reply:
Love this abstraction.